src/server/gc_worker/rawkv_compaction_filter.rs

```diff
@@ -0,0 +1,514 @@
+            stats.filtered.update(|x| x + self.total_filtered);
+            if stats.need_report() {
+                return Some(stats.prepare_report());
+            }
+            None
+        }) {
+            if filtered > 0 {
+                info!("RawKV Compaction filter reports"; "total" => versions, "filtered" => filtered);
+            }
+        }
+    }
+}
+
+pub fn split_ts(key: &[u8]) -> Result<(&[u8], u64), String> {
+    match Key::split_on_ts_for(key) {
+        Ok((key, ts)) => Ok((key, ts.into_inner())),
+        Err(_) => Err(format!(
+            "invalid write cf key: {}",
+            log_wrappers::Value(key)
+        )),
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+
+    use super::*;
+    use api_version::RawValue;
+    use kvproto::kvrpcpb::ApiVersion;
+    use std::thread;
+
+    use crate::config::DbConfig;
+    use crate::server::gc_worker::TestGCRunner;
+    use crate::storage::kv::TestEngineBuilder;
+    use engine_traits::{Peekable, SyncMutable, CF_DEFAULT};
+    use std::time::Duration;
+    use txn_types::TimeStamp;
+
+    pub fn make_key(key: &[u8], ts: u64) -> Vec<u8> {
+        let key1 = Key::from_raw(key)
+            .append_ts(TimeStamp::new(ts))
+            .as_encoded()
+            .to_vec();
+        let res = keys::data_key(key1.as_slice());
+        res
+    }
+
+    #[test]
+    fn test_raw_compaction_filter() {
+        let mut cfg = DbConfig::default();
+        cfg.writecf.disable_auto_compactions = true;
+        cfg.writecf.dynamic_level_bytes = false;
+
+        let engine = TestEngineBuilder::new()
+            .api_version(ApiVersion::V2)
+            .build_with_cfg(&cfg)
+            .unwrap();
+        let raw_engine = engine.get_rocksdb();
+        let mut gc_runner = TestGCRunner::new(0);
+
+        let value1 = RawValue {
+            user_value: vec![0; 10],
+            expire_ts: Some(TimeStamp::max().physical()),
+            is_delete: false,
+        };
+
+        let user_key = b"r\0aaaaaaaaaaa";
+
+        raw_engine
+            .put_cf(
+                CF_DEFAULT,
+                make_key(user_key, 100).as_slice(),
+                &ApiV2::encode_raw_value_owned(value1.clone()),
+            )
+            .unwrap();
+        raw_engine
+            .put_cf(
+                CF_DEFAULT,
+                make_key(user_key, 90).as_slice(),
+                &ApiV2::encode_raw_value_owned(value1.clone()),
+            )
+            .unwrap();
+        raw_engine
+            .put_cf(
+                CF_DEFAULT,
+                make_key(user_key, 70).as_slice(),
+                &ApiV2::encode_raw_value_owned(value1),
+            )
+            .unwrap();
+
+        gc_runner.safe_point(80).gc_raw(&raw_engine);
+        //Wait gc  end
+        thread::sleep(Duration::from_millis(1000));
+
+        // If ts(70) < safepoint(80), and this userkey's latest verion is not deleted or expired, this version will be removed in do_filter.
+        let isexit70 = raw_engine
```

Comment:
What does `isexit` mean ?

Reply:
It should be named is_exist_70.I want to check if the raw(ts=70) still exists